File: leetcode/graph/traversal/897. Increasing Order Search Tree/897. Increasing Order Search Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void inOrder(TreeNode* root, vector <int> &nodes) {
        if(root == NULL) return;
        inOrder(root->left, nodes);
        nodes.push_back(root->val);
        inOrder(root->right, nodes);
    }

    TreeNode* increasingBST(TreeNode* root) {
        vector <int> nodes;
        inOrder(root, nodes);

        TreeNode *ans = new TreeNode();
        TreeNode *cur = ans;
        for(auto node : nodes) {
            TreeNode *newNode = new TreeNode(node);
            cur -> right = newNode;
            cur = cur -> right;
        }
        return ans->right;
    }
};

File: leetcode/graph/traversal/897. Increasing Order Search Tree/897. Increasing Order Search Tree.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;

impl Solution {
    pub fn in_order(root: Option<Rc<RefCell<TreeNode>>>, nodes: &mut Vec<i32>) {
        if let Some(node) = root {
            Self::in_order(node.borrow().left.clone(), nodes);
            nodes.push(node.borrow().val);
            Self::in_order(node.borrow().right.clone(), nodes);
        }
    }

    pub fn increasing_bst(root: Option<Rc<RefCell<TreeNode>>>) -> Option<Rc<RefCell<TreeNode>>> {
        let mut nodes = vec![];
        Self::in_order(root, &mut nodes);
        
        let mut ans = Rc::new(RefCell::new(TreeNode::new(0)));
        let mut cur = ans.clone();

        for node in nodes {
            cur.borrow_mut().right = Some(Rc::new(RefCell::new(TreeNode::new(node))));
            let next = cur.borrow().right.as_ref().unwrap().clone();
            cur = next;
        }

        let ans = ans.borrow().right.clone();
        ans
    }
}
